File: android/src/main/java/com/adyenreactnativesdk/component/base/ModuleException.kt
/*
 * Copyright (c) 2023 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 */

package com.adyenreactnativesdk.component.base

open class KnownException @JvmOverloads constructor(
    val code: String,
    errorMessage: String,
    cause: Throwable? = null
) : RuntimeException(errorMessage, cause)

sealed class ModuleException(code: String, message: String, cause: Throwable? = null) :
    KnownException(code = code, errorMessage = message, cause) {
    class Canceled : ModuleException(
        code = "canceledByShopper",
        message = "Payment canceled by shopper"
    )

    class NotSupported : ModuleException(
        code = "notSupported",
        message = "Not supported on Android"
    )

    class NoClientKey : ModuleException(
        code = "noClientKey",
        message = "No clientKey in configuration"
    )

    class NoPayment : ModuleException(
        code = "noPayment",
        message = "No payment in configuration"
    )

    class InvalidPaymentMethods(e: Throwable?) : ModuleException(
        code = "invalidPaymentMethods",
        message = "Can not parse paymentMethods or the list is empty",
        cause = e
    )

    class InvalidAction(e: Throwable?) : ModuleException(
        code = "invalidAction",
        message = "Can not parse action",
        cause = e
    )

    class NoPaymentMethod(type: String) : ModuleException(
        code = "noPaymentMethod",
        message = "Can not find payment method of type $type in provided list"
    )

    class NoPaymentMethods(type: Collection<String>) : ModuleException(
        code = "noPaymentMethod",
        message = "Can not find payment method of types \"${type.joinToString(",")}\" in provided list"
    )

    class NoModuleListener(integration: String) : ModuleException(
        code = "noModuleListener",
        message = "No DropInService registered for: $integration"
    )

    class Unknown(reason: String?) : ModuleException(
        code = "unknown",
        message = if (reason.isNullOrEmpty()) "Reason unknown" else reason
    )

    class SessionError(error: Throwable?) : ModuleException(
        code = "session",
        message = "Something went wrong while starting session",
        cause = error
    )

    class NoActivity : ModuleException(
        code = "noActivity",
        message = "Launcher not registered. Please call AdyenCheckout.setLauncherActivity() on MainActivity.onCreate()"
    )

    class WrongFlow : ModuleException(
        code = "noActivity",
        message = "ViewModel callback is inconsistent"
    )
}

File: ios/Components/Base/ModuleException.swift
//
// Copyright (c) 2026 Adyen N.V.
//
// This file is open source and available under the MIT license. See the LICENSE file for more info.
//

import Adyen
import Adyen3DS2
import AdyenNetworking
import React
import UIKit

enum ModuleException: LocalizedError, KnownError {
    case canceled
    case noClientKey
    case noPayment
    case notSupported
    case invalidPaymentMethods
    case invalidAction
    case notKeyWindow
    case paymentMethodNotFound(PaymentMethod.Type)
    case balanceCheck(message: String)
    case orderRequest(message: String)
    case sessionError
    case invalidClientKey
    case componentNotRegistered(String)

    var errorCode: String {
        switch self {
        case .canceled:
            return "canceledByShopper"
        case .notSupported:
            return "notSupported"
        case .noClientKey:
            return "noClientKey"
        case .noPayment:
            return "noPayment"
        case .invalidPaymentMethods:
            return "invalidPaymentMethods"
        case .invalidAction:
            return "invalidAction"
        case .paymentMethodNotFound:
            return "noPaymentMethod"
        case .notKeyWindow:
            return "notKeyWindow"
        case .balanceCheck:
            return "balanceCheck"
        case .orderRequest:
            return "orderRequest"
        case .sessionError:
            return "sessionError"
        case .invalidClientKey:
            return "invalidClientKey"
        case .componentNotRegistered:
            return "componentNotRegistered"
        }
    }

    var errorDescription: String? {
        switch self {
        case .canceled:
            return "Payment canceled by shopper"
        case .notSupported:
            return "Not supported on iOS"
        case .noClientKey:
            return "No clientKey in configuration"
        case .invalidClientKey:
            return "Invalid clientKey"
        case .noPayment:
            return "No payment in configuration"
        case .invalidPaymentMethods:
            return "Can not parse paymentMethods or the list is empty"
        case .invalidAction:
            return "Can not parse action"
        case let .paymentMethodNotFound(type):
            return "Can not find payment method of type \(type) in provided list"
        case .notKeyWindow:
            return "Can not find root ViewController"
        case let .balanceCheck(message):
            return "Balance check error: \(message)"
        case let .orderRequest(message):
            return "Order request error: \(message)"
        case .sessionError:
            return "Something went wrong while starting session"
        case let .componentNotRegistered(type):
            return "No embedded component registered for type \(type)"
        }
    }

    static func checkErrorType(_ error: Error) -> Error {
        if error.isComponentCanceled || error.is3DSCanceled {
            return ModuleException.canceled
        }
        if error.isEmpty401NetworkingResponseError {
            return ModuleException.invalidClientKey
        }
        return error
    }
}

extension Error {

    var isComponentCanceled: Bool {
        (self as? CheckoutError)?.code == .cancelled
    }

    var is3DSCanceled: Bool {
        (self as NSError).domain == "com.adyen.Adyen3DS2.ADYRuntimeError" &&
            (self as NSError).code == ADYRuntimeErrorCode.challengeCancelled.rawValue
    }

    var isEmpty401NetworkingResponseError: Bool {
        self is HTTPResponse<EmptyErrorResponse>
    }
}
